File: Kernel Simulator/Shell/Shells/UESH/Commands/KeyInfo.vb

'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Namespace Shell.Shells.UESH.Commands
    Class KeyInfoCommand
        Inherits CommandExecutor
        Implements ICommand

        Public Overrides Sub Execute(StringArgs As String, ListArgs() As String, ListArgsOnly As String(), ListSwitchesOnly As String()) Implements ICommand.Execute
            Write(DoTranslation("Enter a key or a combination of keys to display its information."), True, ColTypes.Neutral)
            Dim KeyPress As ConsoleKeyInfo = Console.ReadKey(True)

            'Pressed key
            Write("- " + DoTranslation("Pressed key") + ": ", False, ColTypes.ListEntry)
            Write(KeyPress.Key.ToString, True, ColTypes.ListValue)

            'If the pressed key is a control key, don't write the actual key char so as not to corrupt the output
            If Not Char.IsControl(KeyPress.KeyChar) Then
                Write("- " + DoTranslation("Pressed key character") + ": ", False, ColTypes.ListEntry)
                Write(KeyPress.KeyChar, True, ColTypes.ListValue)
            End If

            'Pressed key character code
            Write("- " + DoTranslation("Pressed key character code") + ": ", False, ColTypes.ListEntry)
            Write($"0x{Convert.ToInt32(KeyPress.KeyChar):X2} [{Convert.ToInt32(KeyPress.KeyChar)}]", True, ColTypes.ListValue)

            'Pressed modifiers
            Write("- " + DoTranslation("Pressed modifiers") + ": ", False, ColTypes.ListEntry)
            Write(KeyPress.Modifiers.ToString, True, ColTypes.ListValue)

            'Keyboard shortcut
            Write("- " + DoTranslation("Keyboard shortcut") + ": ", False, ColTypes.ListEntry)
            Write($"{String.Join(" +", KeyPress.Modifiers.ToString.Split(", "))} + {KeyPress.Key}", True, ColTypes.ListValue)
        End Sub

    End Class
End Namespace

File: public/Kernel Simulator/Shell/Shells/UESH/Commands/KeyInfo.cs
﻿
// Kernel Simulator  Copyright (C) 2018-2023  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System;
using KS.ConsoleBase.Colors;
using KS.Languages;
using KS.Misc.Writers.ConsoleWriters;
using KS.Shell.ShellBase.Commands;
using KS.ConsoleBase.Inputs;

namespace KS.Shell.Shells.UESH.Commands
{
    /// <summary>
    /// You can view the information about a pressed key
    /// </summary>
    /// <remarks>
    /// This command lets you view the details about a pressed key on your keyboard, including the pressed key and character, the hexadecimal representation of the letter, the pressed modifiers, and the keyboard shortcut.
    /// </remarks>
    class KeyInfoCommand : BaseCommand, ICommand
    {

        public override void Execute(string StringArgs, string[] ListArgsOnly, string[] ListSwitchesOnly)
        {
            TextWriterColor.Write(Translate.DoTranslation("Enter a key or a combination of keys to display its information."));
            var KeyPress = Input.DetectKeypress();

            // Pressed key
            TextWriterColor.Write("- " + Translate.DoTranslation("Pressed key") + ": ", false, KernelColorType.ListEntry);
            TextWriterColor.Write(KeyPress.Key.ToString(), true, KernelColorType.ListValue);

            // If the pressed key is a control key, don't write the actual key char so as not to corrupt the output
            if (!char.IsControl(KeyPress.KeyChar))
            {
                TextWriterColor.Write("- " + Translate.DoTranslation("Pressed key character") + ": ", false, KernelColorType.ListEntry);
                TextWriterColor.Write(Convert.ToString(KeyPress.KeyChar), true, KernelColorType.ListValue);
            }

            // Pressed key character code
            TextWriterColor.Write("- " + Translate.DoTranslation("Pressed key character code") + ": ", false, KernelColorType.ListEntry);
            TextWriterColor.Write($"0x{Convert.ToInt32(KeyPress.KeyChar):X2} [{Convert.ToInt32(KeyPress.KeyChar)}]", true, KernelColorType.ListValue);

            // Pressed modifiers
            TextWriterColor.Write("- " + Translate.DoTranslation("Pressed modifiers") + ": ", false, KernelColorType.ListEntry);
            TextWriterColor.Write(KeyPress.Modifiers.ToString(), true, KernelColorType.ListValue);

            // Keyboard shortcut
            TextWriterColor.Write("- " + Translate.DoTranslation("Keyboard shortcut") + ": ", false, KernelColorType.ListEntry);
            TextWriterColor.Write($"{string.Join(" + ", KeyPress.Modifiers.ToString().Split(new string[] { ", " }, StringSplitOptions.None))} + {KeyPress.Key}", true, KernelColorType.ListValue);
        }

    }
}
